### PR TITLE
[thci] add SSH support for OpenThread wpanctl THCI

### PR DESCRIPTION
**Pull request by FiveDimensions**

### CODE REVIEW

File: tools/harness-thci/OpenThread_WpanCtl.py [hunk 1]
```diff
@@ -192,46 +196,78 @@ def __sendCommand(self, cmd):
...
-                    print "response: %s" % response
-                    if re.match(WPAN_CARRIER_PROMPT, line):
-                        break
-                    elif re.search(r'Not\s+Found|failed\s+with\s+error', line, re.M|re.I):
-                        print "Command failed"
-                        return 'Fail'
+            stdout_lines = []
+            stderr_lines = []
+            if self._is_net:
+
```
Comment (librasungirl) [inline tools/harness-thci/OpenThread_WpanCtl.py]: nit: extra line

File: tools/harness-thci/OpenThread_WpanCtl.py [hunk 2]
```diff
@@ -717,15 +753,19 @@ def _connect(self):
...
+                self.handle.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+                self.handle.connect(self.dutIpv4, port=self.dutPort, username=WPAN_CARRIER_USER, password=WPAN_CARRIER_PASSWD)
+                self.handle.exec_command("stty cols 128")
+                self._is_net = True
+            except Exception, e:
+                ModuleHelper.WriteIntoDebugLogger('connect to ssh Error: ' + str(e))
         else:
             raise Exception('Unknown port schema')
         self.UIStatusMsg = self.getVersionNumber()
-
```
Comment (librasungirl) [inline tools/harness-thci/OpenThread_WpanCtl.py]: keep the blank line before new function?

File: tools/harness-thci/OpenThread_WpanCtl.py [hunk 3]
```diff
@@ -1303,14 +1343,19 @@ def getPANID(self):
         return self.__sendCommand(WPANCTL_CMD + 'getprop -v Network:PANID')[0]
 
     def reset(self):
+        """factory reset"""
```
Comment (librasungirl) [inline tools/harness-thci/OpenThread_WpanCtl.py]: nit: duplicate with below line

**Comment by librasungirl** [comment]
Add description for this new feature in README.md as well?
https://github.com/openthread/openthread/blob/master/tools/harness-thci/README.md

File: tools/harness-thci/OpenThread_WpanCtl.py [hunk 4]
```diff
@@ -69,13 +70,19 @@ def __init__(self, **kwargs):
...
             self.AutoDUTEnable = False
             self._is_net = False                  # whether device is through ser2net
             self.logStatus = {'stop':'stop', 'running':'running', 'pauseReq':'pauseReq', 'paused':'paused'}
             self.joinStatus = {'notstart':'idle', 'ongoing':'discover', 'succeed':'joined', 'failed':'failed'}
             self.logThreadStatus = self.logStatus['stop']
+            self.connectType = (kwargs.get('Param5')).strip().lower() if kwargs.get('Param5') is not None else "usb"
+            if self.connectType == 'ip':
+                self.dutIpv4 = kwargs.get('TelnetIP')
+                self.dutPort = kwargs.get('TelnetPort')
+                self.port = self.dutIpv4 + ":" + self.dutPort
```
Comment (bukepo) [inline tools/harness-thci/OpenThread_WpanCtl.py]: `":"` - > `':'`

File: tools/harness-thci/OpenThread_WpanCtl.py [hunk 5]
```diff
@@ -69,13 +70,19 @@ def __init__(self, **kwargs):
...
             self.UIStatusMsg = ''
             self.mac = kwargs.get('EUI')
-            self.port = kwargs.get('SerialPort')
             self.handle = None
             self.AutoDUTEnable = False
             self._is_net = False                  # whether device is through ser2net
             self.logStatus = {'stop':'stop', 'running':'running', 'pauseReq':'pauseReq', 'paused':'paused'}
             self.joinStatus = {'notstart':'idle', 'ongoing':'discover', 'succeed':'joined', 'failed':'failed'}
             self.logThreadStatus = self.logStatus['stop']
+            self.connectType = (kwargs.get('Param5')).strip().lower() if kwargs.get('Param5') is not None else "usb"
```
Comment (bukepo) [inline tools/harness-thci/OpenThread_WpanCtl.py]: `"usb"` -? `'usb'`

File: tools/harness-thci/OpenThread_WpanCtl.py [hunk 6]
```diff
@@ -192,46 +196,78 @@ def __sendCommand(self, cmd):
...
+                    else:
+                        self._sendline(cmd)
+                        self._expect(cmd)
                 except Exception as e:
                     logging.exception('%s: failed to send command[%s]: %s', self.port, cmd, str(e))
                     if retry_times == 0:
                         raise
-                else:
-                    break
+                    else:
```
Comment (bukepo) [inline tools/harness-thci/OpenThread_WpanCtl.py]: is this change intended?

File: tools/harness-thci/OpenThread_WpanCtl.py [hunk 7]
```diff
@@ -717,15 +753,19 @@ def _connect(self):
...
-            self._is_net = True
+        elif self.connectType == 'ip':
+            print "My IP: %s Port: %s" % (self.dutIpv4, self.dutPort)
+            try:
+                self.handle = paramiko.SSHClient()
+                self.handle.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+                self.handle.connect(self.dutIpv4, port=self.dutPort, username=WPAN_CARRIER_USER, password=WPAN_CARRIER_PASSWD)
+                self.handle.exec_command("stty cols 128")
+                self._is_net = True
+            except Exception, e:
```
Comment (bukepo) [inline tools/harness-thci/OpenThread_WpanCtl.py]: `Exception, e` -> `Exception as e`

**Comment by FiveDimensions** [comment]
@librasungirl  will update README soon

File: tools/harness-thci/README.md [hunk 8]
```diff
@@ -57,3 +57,27 @@ Alternatively, platform developers may follow the instructions below to add a ne
 9. Drag the "OpenThread_WpanCtl: Wpantund + NCP" reference device to the **Test Bed** section with the desired number.
 
 10. Select one or more test cases to execute.
+
+
+The above is for the serial connection mode between the DUT and the Host PC. The ethernet connection (only for SSH) mode is
```
Comment (librasungirl) [inline tools/harness-thci/README.md]: nit: `The ethernet connection (only for SSH) mode is...` -> `The ssh connection (via ip) is ...`
as wifi should also work in theory.
